From 3f38400a83d5fd9848e07685d4b217ff80be93ad Mon Sep 17 00:00:00 2001
From: dianlujitao <dianlujitao@lineageos.org>
Date: Tue, 1 May 2018 18:27:32 +0800
Subject: [PATCH 09/15] UsbDeviceManager: Allow custom boot modes to be treated
 as normal mode

 * AOSP commit dca36e4e loads persistent OEM specific USB functions with
   respect to the current boot mode if it's neither "normal" nor
   "unknown". In this case if the corresponding USB function is not set,
   it'll be set to "none".
 * However, while most bootloaders report boot mode as "normal" after
   reboot, OnePlus 3's bootloader reports "reboot". The system will
   treat it as OEM specific boot mode and try to load specific USB
   functions. As the functions are not set, sys.usb.config will be set
   to none.
 * To address the problem, add a whitelist for OEM specific boot modes,
   and treat them as "normal" mode.

Change-Id: I55bb086e240e8c584cb3f4ff71238a591fbc5037
---
 core/res/res/values/lineage_config.xml              |  4 ++++
 core/res/res/values/lineage_symbols.xml             |  3 +++
 .../com/android/server/usb/UsbDeviceManager.java    | 13 +++++++++++++
 3 files changed, 20 insertions(+)

diff --git a/core/res/res/values/lineage_config.xml b/core/res/res/values/lineage_config.xml
index 9ad5fce5eb0..d573d280003 100644
--- a/core/res/res/values/lineage_config.xml
+++ b/core/res/res/values/lineage_config.xml
@@ -44,4 +44,8 @@
     <!-- Paths to fast charging status file to detect whether an oem fast charger is active -->
     <string-array name="config_oemFastChargerStatusPaths" translatable="false">
     </string-array>
+
+    <!-- The list of boot modes which should be treated as normal mode. -->
+    <string-array name="config_additionalNormalBootModes" translatable="false">
+    </string-array>
 </resources>
diff --git a/core/res/res/values/lineage_symbols.xml b/core/res/res/values/lineage_symbols.xml
index 0ce4b2d6f95..208857a9946 100644
--- a/core/res/res/values/lineage_symbols.xml
+++ b/core/res/res/values/lineage_symbols.xml
@@ -64,4 +64,7 @@
 
     <!-- Path to fast charging status file to detect whether an oem fast charger is active -->
     <java-symbol type="array" name="config_oemFastChargerStatusPaths" />
+
+    <!-- The list of boot modes which should be treated as normal mode. -->
+    <java-symbol type="array" name="config_additionalNormalBootModes" />
 </resources>
diff --git a/services/usb/java/com/android/server/usb/UsbDeviceManager.java b/services/usb/java/com/android/server/usb/UsbDeviceManager.java
index 4f38333f3e4..b763e64a67d 100644
--- a/services/usb/java/com/android/server/usb/UsbDeviceManager.java
+++ b/services/usb/java/com/android/server/usb/UsbDeviceManager.java
@@ -171,6 +171,7 @@ public class UsbDeviceManager {
     private Intent mBroadcastedIntent;
     private boolean mPendingBootBroadcast;
     private static Set<Integer> sBlackListedInterfaces;
+    private static String[] sAdditionalNormalModes;
 
     static {
         sBlackListedInterfaces = new HashSet<>();
@@ -227,6 +228,8 @@ public class UsbDeviceManager {
         mContentResolver = context.getContentResolver();
         PackageManager pm = mContext.getPackageManager();
         mHasUsbAccessory = pm.hasSystemFeature(PackageManager.FEATURE_USB_ACCESSORY);
+        sAdditionalNormalModes = mContext.getResources().getStringArray(
+                com.android.internal.R.array.config_additionalNormalBootModes);
         initRndisAddress();
 
         readOemUsbOverrideConfig();
@@ -411,6 +414,16 @@ public class UsbDeviceManager {
 
     private static boolean isNormalBoot() {
         String bootMode = SystemProperties.get(BOOT_MODE_PROPERTY, "unknown");
+
+        if (sAdditionalNormalModes != null) {
+            for (String mode : sAdditionalNormalModes) {
+                if (mode.equals(bootMode)) {
+                    Slog.d(TAG, "Current boot mode " + bootMode + " treated as normal mode.");
+                    return true;
+                }
+            }
+        }
+
         return bootMode.equals(NORMAL_BOOT) || bootMode.equals("unknown");
     }
 
-- 
2.17.0

